[Doc] Add troubleshooting section to docs (#814)
elastic/apm-agent-dotnet: docs/troubleshooting.asciidoc

ifdef::env-github[]
NOTE: For the best reading experience,
please view this documentation at https://www.elastic.co/guide/en/apm/agent/dotnet[elastic.co]
endif::[]

[[troubleshooting]]

> == Troubleshooting

Author: Yes, the first part of this is copy-paste from Java :)

Something isn't quite working as expected?
Here are some guidelines how to find out what's going wrong.

As a first step, please check if your stack is compatible with the currently <<supported-technologies,supported technologies>>.

Don't worry if you can't figure out what the problem is.
Open a topic in the https://discuss.elastic.co/c/apm[APM discuss forum]
and we will help you out.

IMPORTANT: If you do so, *please attach your debug logs* so that we can analyze the problem.
Upload the *complete* logs to a service like https://gist.github.com.
The logs should include everything from the application startup up until the first request has been executed.

[float]
[[no-data-sent]]
=== No data is sent to the APM Server

If neither errors nor performance metrics are being sent to the APM Server,
it's a good idea to first check your logs and look for output just as the app starts.

In case you don't see anything suspicious in the agent logs (no warning or error) it's recommended to turn the log level to `Trace` for further investigation.

> === Collecting agent logs

Author: Well, this was actually what triggered me to write the whole thing :)

Problem is, people sometimes struggle with collecting logs on ASP.NET classic. As you can see it's not trivial, but I feel this is very much needed in a troubleshooting scenario - otherwise they have no logs. Also, we had lots of questions on how to do this.

I totally agree this is way too complicated. By seeing this I think 1 step we can do mid/long term is to ship `IApmLogger` implementations for popular logging systems as part of the agent. With that the part where we show the `ApmLoggerBridge` (I feel that's the most complicated part) will go away and we just say something like "If you use NLog, just enable it with the code snippet below: (and we put a 1 liner there)"  - and same for Serilog, etc.

So, given that I see lots of users failing in the 1. step of the troubleshooting (collecting logs) I'd keep this and try to simplify this part later by shipping some code that they can just plug in easily. 

What do you think?

> === Collecting agent logs

Reviewer: > Problem is, people sometimes struggle with collecting logs on ASP.NET classic. As you can see it's not trivial, but I feel this is very much needed in a troubleshooting scenario - otherwise they have no logs.

> I'd keep this and try to simplify this part later by shipping some code that they can just plug in easily.

I appreciate the explanation! This makes sense now, and I agree, we should keep it.

The way to collect logs depends on the setup of your application.

==== ASP.NET Core 

In case you activated the agent with the `UseAllElasticApm` or the `UseElasticApm` method as described in <<setup-asp-net-core,the setup section>> the agent will integrate with the https://docs.microsoft.com/en-us/aspnet/core/fundamentals/logging/?view=aspnetcore-3.1[ASP.NET Core logging infrastructure] and it'll log as any other component logs. This means, the agent will pick up the configured logging provider and log into that.

==== ASP.NET Classic

> Unlike ASP.NET Core, ASP.NET (classic) does not have a predefined logging system. In case you already have a logging system in place (like NLog, Serilog, or similar) you can direct the agent logs into your logging system.
> 
> For this you need to create a bridge between the agent's internal logger and your logging system. 

Reviewer: ```suggestion
Unlike ASP.NET Core, ASP.NET (classic) does not have a predefined logging system.
However, if you have a logging system in place, like NLog, Serilog, or similar, you can direct the agent logs into your
logging system by creating a bridge between the agent's internal logger and your logging system. 
```

First implement the `IApmLogger` interface from the `Elastic.Apm.Logging` namespace:

[source,csharp]
----
internal class ApmLoggerBridge : IApmLogger
{
	private readonly Lazy<Logger> _logger;
	public bool IsEnabled(ApmLogLevel level)
	{
		// Check for log level here.
		// Typically you just compare the configured log level of your logger
		// to the input parameter of this method and return if it's the same/higher or not
	}

	public void Log<TState>(ApmLogLevel apmLogLevel, TState state, Exception e, Func<TState, Exception, string> formatter)
	{
		// You can log the given log into your logging system here.
	}
}
----

An example implementation for NLog can be seen https://github.com/elastic/apm-agent-dotnet/blob/master/sample/AspNetFullFrameworkSampleApp/App_Start/ApmLoggerToNLog.cs[in our GitHub repository].

> Then you need to tell the agent to use the `ApmLoggerBridge`.

Reviewer: Remove filler words

```suggestion
Then tell the agent to use the `ApmLoggerBridge`.
```

> For this in ASP.NET Core you need to place the following code into the `Application_Start` method in the `HttpApplication` implementation of your app which is typically in the `Global.asx.cs` file:

Reviewer: I'm confused by this sentence. Are we back to ASP.NET Core? I thought we were still talking about classic?

Author: Good catch, my mistake - this should have been "ASP.NET (classic)". Fixed it.

[source,csharp]
----
AgentDependencies.Logger = new ApmLoggerBridge();
----

> The `AgentDependencies` class lives in the `Elastic.Apm.AspNetFullFramework` namespace and during initialization the agent checks if an additional logger was configured - the agent only does this once, therefore it's important to set it as early in the process as possible (typically in the `Application_Start` method).

Reviewer: ```suggestion
The `AgentDependencies` class lives in the `Elastic.Apm.AspNetFullFramework` namespace.
During initialization, the agent checks if an additional logger was configured--the agent only does this once, so it's important to set it as early in the process as possible (typically in the `Application_Start` method).
```
